Feature: Add X509 options to jwt token policy

strehle opened the pull request:
Abstract:
UAA provides SAML and OIDC for integration to external IdPs.

For SAML the use of X509 (instead of plain / raw keys) is defined in XML-Security. OIDC in UAA is possible with HS256 or RS256 depending on the key material (parameter is signing-key). This hard coded mechanism was fixed with https://github.com/cloudfoundry/uaa/pull/1813. However we need to have options on the token creator side.
Json Web Key standard https://www.rfc-editor.org/rfc/rfc7517 describes the use of X509 in JWT.

Some vendors use X509 for trust, e.g. MS Azure or SAP instead plain keys. Okta supports plain keys.

This PR mainly provides 2 new option parameters for UAA setup ( next to signing-key) :
* signing-alg
  Specify the algorithm, e.g. HS384 or RS512
* signing-cert
  Define a X509 certificate with the signing-key similar to SAML. The Certificate is only a Wrapper for the Public Key, however with the X509 object we can get informations within OIDC which are relevant for trust, e.g. x5t is the thumbprint and used as trust-id. x5c is also helpfuly for other vendors. UAA writes "value" parameter in token_key(s) , e.g. https://login.uaa-acceptance.cf-app.com/token_key but this claim is no standard claim here. x5t , x5c are standard claims and known by other parties. x5c and x5t is needed for Trust Anchor with Azure AD

New in well-known/openid-configuration endpoint is x5t and x5c
The outgoing private_key_jwt authentication will add x5t so that potential integration can be done with X509 trust. e.g.
https://learn.microsoft.com/en-us/azure/active-directory/develop/active-directory-certificate-credentials

strehle committed:
Add options to jwt token keys

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into fix/issue/1816

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into fix/issue/1816

strehle committed:
sonar

strehle committed:
tests

strehle committed:
tests

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into fix/issue/1816

strehle committed:
Add x5t to private_key_jwt authentication

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into fix/issue/1816

strehle committed:
sonar issues

strehle committed:
tests added

strehle committed:
tests added

strehle committed:
tests added

strehle commented:
Sonar

https://sonarcloud.io/summary/new_code?id=cloudfoundry-identity-parent&pullRequest=2116

bruce-ricard commented on model/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKey.java:
```diff
@@ -71,6 +71,9 @@ public enum KeyOperation {
     public static String KTY = "kty";
     public static String ALG = "alg";
 
+    public static final String X5C = "x5c";
```
I know that this is a common pattern in the UAA to have constants named with the exact same string that the string it represents, but I really don't like this pattern. I would like the variables to be named something that helps us understand what the variable is about.

I have no idea what "x5c" or "x5t" means.

For example, please take a look at this [commit](https://github.com/cloudfoundry/uaa/commit/f1450a4b4c3902b2a9b4ab67438ebb22514455fc) I made a while ago. I would like all variables to be named similarly. But please let me know if you disagree.

strehle replied:
The claims are described in https://www.rfc-editor.org/rfc/rfc7517 
They belong to the official parts of Json Web Key Set, what UAA offers in /token_key[s] endpoint
Examples can be found in https://www.rfc-editor.org/rfc/rfc7517#page-25

Should I extend to X509_Thumprint and X509_Certificate instead of x5t and x5c ?

strehle replied:
not needed finally

Tallicia replied:
@jaristiz @strehle and I confirm this is addressed.

bruce-ricard commented on model/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKeyTest.java:
```diff
@@ -0,0 +1,71 @@
+
+public class JsonWebKeyTest {
+
+  private static final String sampleRsaKeys = "{\n" +
```
How would you feel about moving this value to a text file, which the test could load at runtime? I think that it would greatly improve its readability.

strehle replied:
done

Tallicia replied:
@jaristiz @strehle and I confirm this is addressed.

bruce-ricard commented on model/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKeyTest.java:
```diff
@@ -0,0 +1,71 @@
+      assertNotNull(key);
+      assertNotNull(JsonWebKey.getRsaPublicKey(key));
+      assertNotNull(key.getX5t());
+      key.setX5c(null);
```
Are you just trying to test the getters and setters here? If so I think that it could be done in a different, very small test.

If not, I don't understand what this is testing.

strehle replied:
changed

Tallicia replied:
@jaristiz @strehle and I confirm this is addressed.

bruce-ricard commented on model/src/test/java/org/cloudfoundry/identity/uaa/oauth/token/VerificationKeyResponseTest.java:
```diff
@@ -0,0 +1,33 @@
+  }
+
+  @Test
+  void getCertX5c() {
```
I like to try to write each test with 3 sections:

1. setup
2. run of the method or class under test
3. assertions to validate that the run went according to the plan

This test has only one line, one assertion. I am unsure of what it is testing. It looks like we are testing the constructor call from line 21, and maybe the `getCertX5c()` method, or both?

strehle replied:
setupResponse created

Tallicia replied:
@jaristiz @strehle and I confirm this is addressed.

bruce-ricard commented on model/src/test/java/org/cloudfoundry/identity/uaa/zone/TokenPolicyTest.java:
```diff
@@ -88,7 +89,7 @@ public void emptyKeyId() {
     public void deserializationOfTokenPolicyWithVerificationKey_doesNotFail() {
         String jsonTokenPolicy = "{\"keys\":{\"key-id-1\":{\"verificationKey\":\"some-verification-key-1\",\"signingKey\":\"some-signing-key-1\"}}}";
```
Similarly to above, even if this one is significantly shorter than the very long string from above, it is still very hard to read and make sure that it is syntactically correct. 

I would like for these to either be in text files, or maybe constructed using JSON libraries, followed by a `.toString()`.

bruce-ricard replied:
I realize that you didn't write this one, just explaining my thoughts :)

strehle replied:
I changed my part

bruce-ricard commented on model/src/test/java/org/cloudfoundry/identity/uaa/zone/TokenPolicyTest.java:
```diff
@@ -105,7 +106,21 @@ public void tokenPolicy_whenInvalidUniquenessValue_throwsException() {
         TokenPolicy tokenPolicy = JsonUtils.readValue(jsonTokenPolicy, TokenPolicy.class);
-        assertEquals(tokenPolicy.getKeys().get("key-id-1"), "some-signing-key-1");
-        assertEquals(tokenPolicy.getKeys().get("key-id-2"), "some-signing-key-2");
+        assertEquals(tokenPolicy.getKeys().get("key-id-1").getSigningKey(), "some-signing-key-1");
```
I notice that you sometimes have the constant as the first parameter of the assert, and sometimes as the second one.
That has to be incorrect.

I can never remember if the expected value goes first or second in Junit, but I am sure that it should always be in the same position.

strehle replied:
fixed my part

bruce-ricard replied:
It seems to me like some are still backwards. The expected value (the constant) should [come first](https://junit.org/junit4/javadoc/4.8/org/junit/Assert.html#assertEquals(java.lang.Object,%20java.lang.Object)).

Here it should be

```
assertEquals("some-signing-key-1", tokenPolicy.getKeys().get("key-id-1").getSigningKey());
```

bruce-ricard commented on model/src/test/java/org/cloudfoundry/identity/uaa/zone/TokenPolicyTest.java:
```diff
@@ -105,7 +106,21 @@ public void tokenPolicy_whenInvalidUniquenessValue_throwsException() {
+
+    @Test
+    public void initializationWithKeyMap() {
+        String jsonTokenPolicy = "{\"key-id-1\":{\"signingKey\":\"some-signing-key-1\",\"signingCert\":\"some-cert\",\"signingAlg\":\"RS256\"}}";
```
I would like to avoid adding new ones though.

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java:
```diff
@@ -161,11 +174,12 @@ class RsaKeyInfo extends KeyInfo {
 
     public RsaKeyInfo(String keyId, String signingKey, String keyUrl) {
-        this(keyId, signingKey, keyUrl, null);
+        this(keyId, signingKey, keyUrl, null, null);
```
This class is hard to understand with the potential null fields. 

Could we consider maybe extending this class, and adding the optional fields in the child class?

At the very least, I would prefer the nullable fields to be made [Optional](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html) instead.

strehle replied:
added method

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java:
```diff
@@ -285,7 +306,16 @@ public Map<String, Object> getJwkMap() {
+            X509Certificate x509Certificate = X509CertUtils.parse(verifierCertificate);
+            if (x509Certificate != null) {
+                byte[] encoded = JwtHelper.getX509CertEncoded(x509Certificate);
+                result.put("x5c", Collections.singletonList(Base64.encode(encoded).toString()));
```
Shouldn't we use the "x5c" and "x5t" constants here?

strehle replied:
we can do but then we should refactor all usages, e.g. kid, kty etc

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java:
```diff
@@ -285,7 +306,16 @@ public Map<String, Object> getJwkMap() {
+                byte[] encoded = JwtHelper.getX509CertEncoded(x509Certificate);
+                result.put("x5c", Collections.singletonList(Base64.encode(encoded).toString()));
+                result.put("x5t", JwtHelper.getX509CertThumbprint(encoded, "SHA-1"));
+                result.put("x5t#S256", JwtHelper.getX509CertThumbprint(encoded, "SHA-256"));
```
I don't understand what `x5t#S256` is.

strehle replied:
hash with sha-256 instead sha-1 , see rfc 7517

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfoBuilder.java:
```diff
@@ -6,9 +6,9 @@
 
     public static KeyInfo build(String keyId, String signingKey, String uaaUrl) {
-        return build(keyId, signingKey, uaaUrl, null);
+        return build(keyId, signingKey, uaaUrl, null, null);
```
I would prefer to have a method override that would prevent us from passing 2 `nulls` in here, it makes the code hard to read.

strehle replied:
done

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtHelper.java:
```diff
@@ -55,13 +59,42 @@ public static Jwt decode(String token) {
+    public static byte[] getX509CertEncoded(X509Certificate x509Certificate) {
+        try {
+            return x509Certificate.getEncoded();
+        } catch (Exception e) {
```
I am upset that the static analysis doesn't catch `catch (Exception ...`. We should catch `RuntimeException`, to prevent the code from potentially catching checked exceptions in the future. Never `Exception`.

strehle replied:
done

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtHelper.java:
```diff
@@ -55,13 +59,42 @@ public static Jwt decode(String token) {
+        try {
+            MessageDigest sha256 = MessageDigest.getInstance(alg);
+            return Base64URL.encode(sha256.digest(derEncodedCert)).toString();
+        } catch (Exception e) {
```
Same thing here.

strehle replied:
done

bruce-ricard commented on server/src/test/java/org/cloudfoundry/identity/uaa/impl/config/LegacyTokenKeyTest.java:
```diff
@@ -14,7 +14,7 @@ public class LegacyTokenKeyTest {
     @Test
     public void shouldBuildLegacyTokenKey_withSecureKeyUrl() {
-        LegacyTokenKey.setLegacySigningKey("secret", "http://uaa.url");
+        LegacyTokenKey.setLegacySigningKey("secret", "http://uaa.url", null, null);
```
I would really prefer not having `null, null` in the parameter list.

bruce-ricard commented on server/src/test/java/org/cloudfoundry/identity/uaa/impl/config/UaaConfigurationTokenKeyTest.java:
```diff
@@ -0,0 +1,32 @@
+
+  @Test
+  void testTokenKey() {
+    assertNotNull(createValidator("uaa:\n" +
```
I think that's another example of something that would fix better in a text file

strehle replied:
moved to file

Tallicia replied:
@strehle and I agree.

bruce-ricard commented on server/src/test/java/org/cloudfoundry/identity/uaa/impl/config/UaaConfigurationTokenKeyTest.java:
```diff
@@ -0,0 +1,32 @@
+
+  @Test
+  void testTokenKey() {
+    assertNotNull(createValidator("uaa:\n" +
```
It doesn't feel right to have this large input to the `createValidator` function, and the only thing we test is that the output is not null. Could we test a little more?

strehle replied:
This part is outdated

Tallicia replied:
@strehle and I confirmed this is adjusted.

bruce-ricard commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtHelperTest.java:
```diff
@@ -31,4 +31,10 @@ public void jwtHeaderShouldContainJkuInTheHeader() {
+    @Test
+    public void jwtHeaderShouldContainJkuInTheHeaderEvenWithX5t() {
+        Jwt jwt = JwtHelper.encode("testJwtContent", keyInfo, true);
+        assertThat(jwt.getHeader().getJku(), is("https://localhost/uaa/token_keys"));
```
Shouldn't we try to use only one assert format consistently throughout the codebase?

i.e. either use the `assertEquals(x,y)` or this `assertThat(x, is(y))`

strehle replied:
This part is outdated

Tallicia replied:
@strehle and I confirmed this is adjusted.

bruce-ricard commented on server/src/test/java/org/cloudfoundry/identity/uaa/zone/GeneralIdentityZoneConfigurationValidatorTests.java:
```diff
@@ -358,4 +360,18 @@ public void validate_invalid_corsPolicy_defaultConfiguration_allowedOrigins() th
+        zone.getConfig().getTokenPolicy().setKeyInformation(keyInformationMap);
+        zone.getConfig().getTokenPolicy().setActiveKeyId("id-1");
+
+        validator.validate(zone, mode);
```
I'm guessing that the test is testing that this call doesn't throw an exception. 

If that's the case, how would you feel about using [assertDoesNotThrow](https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assertions.html#assertDoesNotThrow(org.junit.jupiter.api.function.Executable)), I think that it makes the test more readable.

strehle replied:
This part is outdated

Tallicia replied:
@strehle and I confirmed this is adjusted.

bruce-ricard reviewed (changes requested):
Hi Markus. I did a first pass code review. I still don't fully understand the feature, but I made some readability comments inline. Please let me know what you think.

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into fix/issue/1816

strehle committed:
review

strehle committed:
review

strehle committed:
review

strehle committed:
review

strehle committed:
review

strehle committed:
test fix

strehle commented:
we can refactor constants if this helps, see https://github.com/cloudfoundry/uaa/pull/2181

strehle committed:
use constants from nimbus jose

bruce-ricard commented on model/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKey.java:
```diff
@@ -71,8 +72,8 @@ public enum KeyOperation {
-    public static final String X5C = "x5c";
-    public static final String X5T = "x5t";
+    public static final String X5C = HeaderParameterNames.X_509_CERT_CHAIN;
+    public static final String X5T = HeaderParameterNames.X_509_CERT_SHA_1_THUMBPRINT;
```
I like this library. Did you consider not creating your own constants at all, but just using the library provided ones?

It seems to me like we don't need `X5C` and `X5T` anymore.

strehle replied:
removed

Tallicia replied:
@strehle and I confirmed this has been adjusted.

bruce-ricard commented on model/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKeyTest.java:
```diff
@@ -4,56 +4,23 @@
+  private static final String microsoftJwKSet = getResourceAsString(JsonWebKeyTest.class, "JwkSet-Microsoft.json");
+  // UAA jwks_uri: https://login.uaa-acceptance.cf-app.com/token_keys
+  private static final String uaaLegacyJwkSet = getResourceAsString(JsonWebKeyTest.class, "JwkSet-LegacyUaa.json");
+  // https://www.keycloak.org/ jwks_uri: https://ssotest.domain.de/auth/realms/tenantkey/protocol/openid-connect/certs
```
I'm getting:
>ssotest.domain.de’s server IP address could not be found

That appears to be a dead link

strehle replied:
changed to Keycloak server documentation.

strehle replied:
removed these tests, therefore obsolete

bruce-ricard commented on model/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKeyTest.java:
```diff
@@ -4,56 +4,23 @@
-      "        }\n" +
-      "    ]\n" +
-      "}";
+  // Azure AD jwks_uri : https://login.microsoftonline.com/9bc40aaf-e150-4c30-bb3c-a8b3b677266e/discovery/v2.0/keys
```
I don't understand what these URLs are. Is that where you got the json from?

strehle replied:
These are examples for different jwks_uri from various vendors/products. jwks_uri and content is defined in RFC 7517, but there are optionals and therefore each product has different structure

bruce-ricard commented on model/src/test/resources/org/cloudfoundry/identity/uaa/oauth/jwk/JwkSet-Keycloak.json:
```diff
@@ -0,0 +1,17 @@
+{
```
This is great IMO. So much easier to read the value. Thanks!

Tallicia replied:
@strehle and I agree.

Tallicia replied:
@strehle and I agree.

bruce-ricard reviewed (changes requested):
Thanks for the fixes, looking good!

Requesting changes mostly for the dead link.

Apologies for now reviewing earlier, we've been really busy.

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into fix/issue/1816

strehle committed:
review

strehle committed:
Added documentation

strehle commented:
@bruce-ricard  any update regarding review.
This PR would solve 2 issue 
- private_key trust to Azure OIDC IDP
- Support PS256, in order to be complaint with newer standards https://openid.net/specs/openid-financial-api-ciba.html#algorithm-considerations

Solved support ES256 signature support with https://github.com/cloudfoundry/uaa/pull/2208
moved some refactorings and tests into https://github.com/cloudfoundry/uaa/pull/2208 so that it should get clearer now what this PR should add

bruce-ricard commented on model/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKeyTest.java:
```diff
@@ -0,0 +1,78 @@
+import static org.junit.Assert.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+
+public class JsonWebKeyTest {
```
I am having a hard time understanding *what* these tests are actually testing.

It appears to us that a lot of these tests are testing that the json is properly deserialized, meaning that we are testing the `JsonWebKeyDeserializer` class, that is called through the `JsonUtils.readValue` function. If that is indeed the case, shouldn't these tests be in a file called `JsonWebKeyDeserializerTest` instead of this one?

Tallicia replied:
@jaristiz @strehle and I confirm this is addressed.

bruce-ricard commented on model/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKeyTest.java:
```diff
@@ -0,0 +1,78 @@
+  private static final String keyCloakJwkSet = getResourceAsString(JsonWebKeyTest.class, "JwkSet-Keycloak.json");
+
+  @Test
+  public void testWebKeysMicrosoft() {
```
Typically, this test appears to be testing only the deserialization.

strehle replied:
removed these method and testWebKeysMicrosoft

bruce-ricard commented on model/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKeyTest.java:
```diff
@@ -0,0 +1,78 @@
+    for (JsonWebKey key : keys.getKeys()) {
+      assertNotNull(key);
+      assertNotNull(JsonWebKey.getRsaPublicKey(key));
+      assertNotNull(key.getX5t());
```
I would like to see more actual equality assertions. It seems to me like here we should be able to assert that the `X5t` is the actual value that is stored in the JSON file. 

Just asserting that it's not null could yield a false negative.

Tallicia replied:
@jaristiz @strehle and I confirm this is addressed.

bruce-ricard commented on model/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKeyTest.java:
```diff
@@ -0,0 +1,78 @@
+    });
+    JsonWebKeySet<JsonWebKey> keys = new JsonWebKeySet<>(uaaKeys.getKeys());
+    keys.getKeys().forEach(k -> k.setX5t("x509Thumbprint"));
+    keys.getKeys().forEach(k -> k.setX5c(new String[] { "x509-Certificate" } ));
```
These 2 lines confuse me. I don't understand what you are trying to achieve by setting these 2 values here, and asserting later that they were correctly set.

Are you testing that the setters and getters are working as expected?

Tallicia replied:
@strehle and I confirm this is addressed and corrected.

bruce-ricard commented on model/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKeyTest.java:
```diff
@@ -0,0 +1,78 @@
+    for (JsonWebKey key : keys.getKeys()) {
+      assertNotNull(key);
+      assertEquals("x509Thumbprint", key.getX5t());
+      assertEquals("x509-Certificate", key.getX5c()[0]);
```
These are the assertions I'm talking about in the above comment.

strehle replied:
moved

Tallicia replied:
@strehle and I confirm this is addressed and corrected.

bruce-ricard commented on model/src/test/java/org/cloudfoundry/identity/uaa/zone/TokenPolicyTest.java:
```diff
@@ -105,7 +106,21 @@ public void tokenPolicy_whenInvalidUniquenessValue_throwsException() {
+        final String sampleIdentityZone = getResourceAsString(getClass(), "SampleIdentityZone.json");
+        IdentityZone identityZone = JsonUtils.readValue(sampleIdentityZone, IdentityZone.class);
+        TokenPolicy tokenPolicy = identityZone.getConfig().getTokenPolicy();
+        assertEquals("some-signing-key-1", tokenPolicy.getKeys().get("key-id-1").getSigningKey());
+        assertEquals("some-cert", tokenPolicy.getKeys().get("key-id-1").getSigningCert());
+        assertEquals("RS256", tokenPolicy.getKeys().get("key-id-1").getSigningAlg());
+        tokenPolicy.setKeys(null);
+        assertEquals("some-signing-key-1", tokenPolicy.getKeys().get("key-id-1").getSigningKey());
+        assertEquals("some-cert", tokenPolicy.getKeys().get("key-id-1").getSigningCert());
+        assertEquals("RS256", tokenPolicy.getKeys().get("key-id-1").getSigningAlg());
```
These seem to all have the expected value first, which is good.

Tallicia replied:
@strehle and I agree.

bruce-ricard commented on model/src/test/java/org/cloudfoundry/identity/uaa/zone/TokenPolicyTest.java:
```diff
@@ -105,7 +106,21 @@ public void tokenPolicy_whenInvalidUniquenessValue_throwsException() {
-        assertEquals(tokenPolicy.getKeys().get("key-id-1"), "some-signing-key-1");
-        assertEquals(tokenPolicy.getKeys().get("key-id-2"), "some-signing-key-2");
+        assertEquals(tokenPolicy.getKeys().get("key-id-1").getSigningKey(), "some-signing-key-1");
+        assertEquals(tokenPolicy.getKeys().get("key-id-2").getSigningKey(), "some-signing-key-2");
```
This one is also backwards.

Tallicia replied:
@strehle and I confirm this is addressed and corrected.

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java:
```diff
@@ -84,6 +92,7 @@ public HmacKeyInfo(String keyId, String signingKey, String keyUrl, String sigAlg
         SecretKey hmacKey = new SecretKeySpec(signingKey.getBytes(), algorithm);
         this.signer = new MacSigner(algorithm, hmacKey);
         this.verifier = new MacSigner(algorithm, hmacKey);
+        this.verifierCertificate = null;
```
As it appears that it is possible for the certificate to have no value, and it to be a valid value, I would prefer for the field to be of the type `Optional<String>` ([Optional](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html) class), to avoid having `null`s and thus preventing potential `NullPointerException`s.

Then this line would become
`this.verifierCertificate = Optional.empty()`
[Doc](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#empty--)

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java:
```diff
@@ -177,6 +192,7 @@ public RsaKeyInfo(String keyId, String signingKey, String keyUrl, String sigAlg)
         this.verifier = new RsaVerifier(rsaPublicKey, algorithm);
         this.keyId = keyId;
         this.verifierKey = pemEncodePublicKey;
+        this.verifierCertificate = signingCert;
```
and this line would become

```
this.verifierCertificate = Optional.of(signingCert);
```

https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#of-T-

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java:
```diff
@@ -285,7 +307,16 @@ public Map<String, Object> getJwkMap() {
         result.put("kty", RSA.name());
-
+        // X509 releated values from JWK spec
+        if (this.verifierCertificate != null) {
```
And here we could use [ifPresent()](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#ifPresent-java.util.function.Consumer-).

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthentication.java:
```diff
@@ -35,7 +35,7 @@ public String getClientAssertion(OIDCIdentityProviderDefinition config) {
     claims.setIat((int) Instant.now().minusSeconds(120).getEpochSecond());
     claims.setExp(Instant.now().plusSeconds(300).getEpochSecond());
-    return JwtHelper.encode(JsonUtils.writeValueAsString(claims), keyInfoService.getActiveKey()).getEncoded();
+    return JwtHelper.encode(JsonUtils.writeValueAsString(claims), keyInfoService.getActiveKey(), true).getEncoded();
```
It's really hard (impossible?) when reading this line to understand what `true` means. This is one of the reasons why I personally try very hard not to add flag arguments to functions.

Clean code says:
> Flag arguments are ugly. Passing a boolean into a function is a truly terrible practice. It
immediately complicates the signature of the method, loudly proclaiming that this function
does more than one thing. It does one thing if the flag is true and another if the flag is false!

In the future, maybe consider having 2 functions. One called `encodeWithoutX5t` and `encodeWithX5t`. It's IMHO a lot easier to read, and it makes both functions simpler and smaller.

strehle replied:
seperated into
* encode
* encodePlusX5t

https://github.com/cloudfoundry/uaa/blob/fix/issue/1816/server/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthentication.java#L39-L40

Tallicia replied:
Bruce has approved, I am resolving this conversation.

bruce-ricard commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtHelperTest.java:
```diff
@@ -29,6 +32,14 @@ public void testKidInHeader() {
+        KeyInfo rsaKeyInfo = KeyInfoBuilder.build("key-id-1", privatekey, "http://localhost/uaa", "RS256", certificate);
+        Jwt jwt = JwtHelper.encode("testJwtContent", rsaKeyInfo, true);
+        assertNull(jwt.getHeader().getJku());
+        assertEquals(jwt.getHeader().getX5t(), "8aFXmG4WA6wAiViW2DF2E6uigBU");
```
This assertion also appears to be backwards.

strehle replied:
done

Tallicia replied:
@strehle and I confirm this is addressed and corrected.

bruce-ricard commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtHelperX5tTest.java:
```diff
@@ -0,0 +1,86 @@
+  @Test
+  public void jwtHeaderShouldContainX5tInTheHeader() {
+    Jwt jwt = JwtHelper.encode("testJwtContent", keyInfo, true);
+    assertThat(jwt.getHeader().getX5t(), is("ijN2hCBB85pSpHSUQGBLK2xGurY"));
```
I like this format. It's very readable IMO with the `assertThat(this value, is(this other one))`. 

Maybe that's why you got the other backwards? Because you are thinking of this format?

strehle replied:
done

Tallicia replied:
@strehle and I confirm this is addressed and corrected.

bruce-ricard commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/token/KeyInfoServiceTests.java:
```diff
@@ -48,7 +48,7 @@ class KeyInfoServiceTests {
     @BeforeAll
     static void setupLegacyKey() {
-        LegacyTokenKey.setLegacySigningKey("testLegacyKey", "https://localhost/uaa");
+        LegacyTokenKey.setLegacySigningKey("testLegacyKey", "https://localhost/uaa", null, null);
```
It looks like to could use the other constructor here, no?

strehle replied:
KeyInfo should be fixed, yes, but this should be done with https://github.com/cloudfoundry/uaa/pull/2208

Tallicia replied:
I'm leaning toward a seperate PR to address additionals.

strehle replied:
Handle Optional for both internally, means sigAlg and certificate but here stay with null if not available.

Tallicia replied:
Bruce has approved, I am resolving this conversation.

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java:
```diff
@@ -161,11 +175,12 @@ class RsaKeyInfo extends KeyInfo {
 
     public RsaKeyInfo(String keyId, String signingKey, String keyUrl) {
-        this(keyId, signingKey, keyUrl, null);
+        this(keyId, signingKey, keyUrl, null, null);
```
It would really like to avoid `null`s. Here too, replacing the nulls with `Optional.empty()` would make the code a lot sturdier, by preventing NPEs.

strehle replied:
with https://github.com/cloudfoundry/uaa/pull/2208 we can refactor the other null, however from experieces I have Optional does not really solve it because now you know xxx != null is needed or ObjectsUtils.isNull / isEmpty.
.... with Optinal you need to be carefully with the .get()
So I have not see less Nullpointer expections only moving to Optinal, but we can do the refactor.

strehle replied:
Use internally Optional and assign it with Optional.ofNullable(signingCert);

Tallicia replied:
Bruce has approved, I am resolving this conversation.

bruce-ricard reviewed (changes requested):
A few more technical comments. 

Let me know what you think.

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into fix/issue/1816

strehle committed:
rebase

strehle committed:
rename

strehle committed:
review

strehle committed:
refactorings

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into fix/issue/1816

strehle committed:
add X509 validation check

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into fix/issue/1816

Tallicia commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtHelperX5tTest.java:
```diff
@@ -0,0 +1,86 @@
+      "ZLzGcSYwCmj0Vc0GO5NgnFVZg4N9CyYCpDMeQynumlrNhRgnZRzlqXtQgL2bQDiu\n" +
+      "coxNL/KY05iVlE1bmq/fzNEmEi2zf3dQV8CNSYs=\n" +
+      "-----END CERTIFICATE-----";
+
```
Moving this to file would be ideal.

strehle replied:
fixed with getResourceAsString to load keys

Tallicia reviewed (commented):
Updated and confirmed resolutions where applicable. A few items remain.

Tallicia reviewed (changes requested):
We need to align on keeping the current codebase consistent and offer the use of optional in a subsequent refactor PR or phased refactoring.

strehle committed:
solved smells from sonar

https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&sinceLeakPeriod=true&pullRequest=2116&id=cloudfoundry-identity-parent

strehle committed:
solved smells from sonar

https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&sinceLeakPeriod=true&pullRequest=2116&id=cloudfoundry-identity-parent

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into fix/issue/1816

strehle committed:
test fix

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into fix/issue/1816

strehle committed:
sonar

strehle committed:
use Optional for verifierCertificate in KeyInfo

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into fix/issue/1816

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into fix/issue/1816
